src/deploy-cloudrun.ts

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as core from '@actions/core';
+import * as exec from '@actions/exec';
+import * as setupGcloud from '../setup-google-cloud-sdk/src/';
+
+/**
+ * Executes the main action. It includes the main business logic and is the
+ * primary entry point. It is documented inline.
+ */
+export async function run(): Promise<void> {
+  try {
+    // Get inputs
+    // Core inputs
+    const image = core.getInput('image'); // Image ie gcr.io/...
+    const name = core.getInput('service'); // Service name
+    const metadata = core.getInput('metadata'); // YAML file
+    const credentials = core.getInput('credentials'); // Service account key
+    let projectId = core.getInput('project_id');
+    // Flags
+    const envVars = core.getInput('env_vars'); // String of env vars KEY=VALUE,...
+    const region = core.getInput('region') || 'us-central1';
+
+    let installBeta = false; // Flag for installing gcloud beta components
+    let cmd;
+
+    // Find base command
+    if (metadata) {
+      // Deploy service from metadata
+      if (image || name || envVars) {
+        core.warning(
+          'Metadata YAML provided: ignoring `image`, `service`, and `env_vars` inputs.',
+        );
+      }
+      cmd = [
+        'beta',
+        'run',
+        'services',
+        'replace',
+        metadata,
+        '--platform',
+        'managed',
+        '--region',
+        region,
+      ];
+      installBeta = true;
+    } else {
+      // Deploy service with image specified
+      cmd = [
+        'run',
+        'deploy',
+        name,
+        '--image',
+        image,
+        '--quiet',
+        '--platform',
+        'managed',
+        '--region',
+        region,
+      ];
+    }
+    if (!metadata) {
+      // Set optional flags from inputs
+      if (envVars) cmd.push('--update-env-vars', envVars);
+    }
+
+    // Install gcloud if not already installed.
+    if (!setupGcloud.isInstalled()) {
```

Comment:
Should give users an option to pin gcloud version for idempotent runs?

Reply:
sg